### PR TITLE
Automate releases

### PR DESCRIPTION
**Pull request by k0da**
* Adds separate Dockerfile for each architecture to keep base image
      unchanged.
    * Brings github_changelog_generator into a release pipeline
    * Add .goreleaser.yaml with go build and amd64/arm64 images with docker
      common manifest
    * Introduce pipeline for tag event, where docker artifacts will be
      built.
    * Rename helm pipeline (release) into helm_publish

Signed-off-by: Dinar Valeev <dinar.valeev@absa.africa>

### CODE REVIEW

**Comment by ytsarev** [comment]
@somaritane keen to release `v0.7.6` using this one?

File: .goreleaser.yml
Comment (ytsarev) [inline .goreleaser.yml:43]: Why exclude them?

File: .goreleaser.yml
Comment (ytsarev) [inline .goreleaser.yml:32]: Will the final tag include `$arch` suffix?

If yes, it will break this logic https://github.com/AbsaOSS/k8gb/blob/master/chart/k8gb/templates/operator.yaml#L29

File: .goreleaser.yml
Comment (ytsarev) [inline .goreleaser.yml:37]: Can you elaborate on the logic here?

**Review by ytsarev — COMMENTED** [review]
Looks super fancy, just some clarifications required

File: .goreleaser.yml
Comment (k0da) [reply .goreleaser.yml:43]: This is upstream behaviour and ignored by providing --release-notes flag, https://goreleaser.com/customization/release/#custom-release-notes

**Comment by somaritane** [comment]
> @somaritane keen to release `v0.7.6` using this one? 

@ytsarev, Oops, I've just released `v0.7.6` now, but `v0.7.7` - definitely :)

File: .goreleaser.yml
Comment (k0da) [reply .goreleaser.yml:32]: It follows same strategy as of now. We build two images imag-{.Arch} then wrap it with single manifest name_template: absaoss/k8gb:{{ .Tag }}

File: .goreleaser.yml
Comment (k0da) [reply .goreleaser.yml:37]: you can run goreleaser --snapshot, it will build image: image:tag-next and publish it

Comment (k0da) [reply .goreleaser.yml:37]: I was wrong on publish: https://goreleaser.com/customization/snapshots/#how-it-works

File: .goreleaser.yml
Comment (k0da) [reply .goreleaser.yml:32]: So it won't break ``Deployment`` manifest. absaoss/k8gb:{{ .Tag }} will resolve into absaoss/k8gb:{{ .Tag }}-amd64 if kubernetes worker is amd64

File: .goreleaser.yml
Comment (ytsarev) [reply .goreleaser.yml:43]: I get it, just don't think we should exclude docs or tests from the changelog :)

File: .goreleaser.yml
Comment (ytsarev) [reply .goreleaser.yml:32]: Cool!

**Comment by ytsarev** [comment]
@k0da will it create Draft Release or Release right away?

**Comment by k0da** [comment]
# If set to true, will not auto-publish the release.
  # Default is false.
  # draft: false
Release straighaway. I think we should start with draft first

**Comment by somaritane** [comment]
>  I think we should start with draft first
Makes sense. `github-changelog-generator` uses GH issue's tags for changelog generation, which might required 2nd pair of eyes to check that issues are correctly categorized

**Comment by k0da** [comment]
amended

File: .goreleaser.yml
Comment (k0da) [reply .goreleaser.yml:39]: Yes, I'll drop it.

**Comment by k0da** [comment]
> @k0da what about we create some fake releases and see how it works end-to-end?

I can tag local commit with v0.7.7-alpha1 and run releaser manually?

**Review by ytsarev — DISMISSED** [review]
https://github.com/AbsaOSS/k8gb/releases/tag/untagged-d1d7bd1b09241abe6eb0 looks great, so let's accommodate the cool pipes :)

**Review by somaritane — DISMISSED** [review]
lgtm

File: .gitignore
Comment (ytsarev) [inline .gitignore:114]: How do we update `CHANGELOG.md` in this scenario?

Comment (k0da) [reply .gitignore:114]: same way as before, post release commit (post release github_changelog_generator on maintainers machine and open PR). 
During GH run only temporary diff is generated into gitingored file.

Comment (ytsarev) [reply .gitignore:114]: But before local github_changelog_generator diff was the input for release. Now it will be required to run it locally post release and commit? It might lead to deviations

Comment (k0da) [reply .gitignore:114]: If I look at last v0.7.6 release, offline changelog was commited at 18:00, and release was done on 15:06. I see no deviation.

Comment (ytsarev) [reply .gitignore:114]: yeas, because offline changelog is the source for release text

Comment (k0da) [reply .gitignore:114]: Is it worth to sacrafice automation in order to have "offline" change log?
you have two tags, and tool to work with those tags. ATM we run tool get output and paste it into "Release".
I propose to run tool as part of release draft pipeline to format a release. Then run tool with same input to commit post release offline changelog. Sorry I only see an improvement here.

Additionally draft can be changed the same way from "offline" changelog.

Comment (ytsarev) [reply .gitignore:114]: I am not talking about sticking to current scenario, more thinking of avoiding pitfalls with the new one. Can we can write to CHANGELOG.md in post-release pipeline and commit it automatically?

Comment (k0da) [reply .gitignore:114]: This option can be explored for sure.

Comment (k0da) [reply .gitignore:114]: @ytsarev ammended

Comment (ytsarev) [reply .gitignore:114]: @k0da very cool, thanks!

File: .github/workflows/helm_publish.yaml
Comment (ytsarev) [inline .github/workflows/helm_publish.yaml:12]: Please switch to v2

**Review by ytsarev — CHANGES_REQUESTED** [review]

File: .github/workflows/helm_publish.yaml
Comment (k0da) [reply .github/workflows/helm_publish.yaml]: done

**Comment by k0da** [comment]
so I tested pull-request action here https://github.com/k0da/terraform-controller/runs/2020895388?check_suite_focus=true
results into https://github.com/k0da/terraform-controller/pull/3

**Comment by k0da** [comment]
IMO any way automated PR can be closed any time. So this is quite harmless improvement

**Review by ytsarev — APPROVED** [review]
lgtm

**Review by somaritane — APPROVED** [review]
lgtm, can't wait for new release :)🤞

**Comment by k0da** [comment]
@somaritane @ytsarev I've ammended fetch-depth param to checkout action to preserve full history for github_changelog_generator

**Review by ytsarev — APPROVED** [review]

**Review by somaritane — APPROVED** [review]

**Pull request merged**
